crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -232,44 +241,94 @@ impl CumulocityConverter {
             mqtt_publisher,
             service_type,
             c8y_endpoint,
+            entity_store,
         })
     }
 
+    /// Performs auto-registration process for an entity under a given
+    /// identifier.
+    ///
+    /// If an entity is a service, its device is also auto-registered if it's
+    /// not already registered.
+    ///
+    /// It returns MQTT register messages for the given entities to be published
+    /// by the mapper, so other components can also be aware of a new device
+    /// being registered.
+    fn auto_register_entity(&mut self, entity_id: &str) -> Vec<Message> {
+        let mut register_messages = vec![];
+        let (device_id, service_id) = match entity_id.split('/').collect::<Vec<&str>>()[..] {
+            ["device", device_id, "service", service_id, ..] => (device_id, Some(service_id)),
+            ["device", device_id, "", ""] => (device_id, None),
+            _ => return register_messages,
+        };
+
+        // register device if not registered
+        let device_topic = format!("{MQTT_ROOT}/device/{device_id}//");
+        if self.entity_store.get(&device_topic).is_none() {
+            let device_type = if device_id == "main" {
+                "device"
+            } else {
+                "child-device"
+            };
+            let device_name = if device_id == "main" {
+                self.entity_store.main_device_name()
+            } else {
+                device_id
+            };
+            let device_register_payload =
+                format!("{{ \"@type\":\"{device_type}\", \"@id\":\"{device_name}\"}}");
```

Comment:
Compared to the original version https://github.com/thin-edge/thin-edge.io/pull/2169/files#diff-1d1eb076015e0658044f7b03b0b805fce4f0253f4048587797fc5dd42019f908L149

- the device type is now correct for the main device.
- the device name is compatible with the current version of thin-edge (to avoid breaking the tests)